File: app/src/main/java/com/example/c001apk/view/LinearAdapterLayout.java
package com.example.c001apk.view;

import android.content.Context;
import android.database.DataSetObserver;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.LinearLayout;

import java.util.ArrayList;

public class LinearAdapterLayout extends LinearLayout {
    private BaseAdapter mAdapter;
    private DataSetObserver mObserver;

    public LinearAdapterLayout(Context context, AttributeSet attributeSet) {
        super(context, attributeSet);
    }

    public static void updateChildView1(LinearAdapterLayout linearAdapterLayout) {
        linearAdapterLayout.updateChildView();
    }

    public BaseAdapter getAdapter() {
        return this.mAdapter;
    }

    public void setAdapter(BaseAdapter baseAdapter) {
        if (this.mAdapter == baseAdapter) {
            return;
        }
        if (this.mObserver == null) {
            this.mObserver = new LinearAdapterLayoutObserver(this);
        }
        BaseAdapter baseAdapter2 = this.mAdapter;
        if (baseAdapter2 != null) {
            baseAdapter2.unregisterDataSetObserver(this.mObserver);
        }
        this.mAdapter = baseAdapter;
        BaseAdapter baseAdapter3 = this.mAdapter;
        if (baseAdapter3 != null) {
            baseAdapter3.registerDataSetObserver(this.mObserver);
        }
        updateChildView();
    }

    public void notifyDataSetChange() {
        updateChildView();
    }

    public void updateChildView() {
        if (this.mAdapter == null) {
            removeAllViews();
            return;
        }
        ArrayList arrayList = new ArrayList();
        for (int i = 0; i < this.mAdapter.getCount(); i++) {
            arrayList.add(this.mAdapter.getView(i, getChildAt(i), this));
        }
        removeAllViewsInLayout();
        for (int i2 = 0; i2 < arrayList.size(); i2++) {
            ViewGroup.LayoutParams layoutParams = ((View) arrayList.get(i2)).getLayoutParams();
            if (layoutParams == null) {
                layoutParams = generateDefaultLayoutParams();
            }
            addViewInLayout((View) arrayList.get(i2), i2, layoutParams);
        }
        requestLayout();
        invalidate();
    }


}

File: app/src/main/java/com/example/c001apk/view/LinearAdapterLayout.kt
package com.example.c001apk.view

import android.content.Context
import android.database.DataSetObserver
import android.util.AttributeSet
import android.view.View
import android.widget.BaseAdapter
import android.widget.LinearLayout

class LinearAdapterLayout(context: Context?, attributeSet: AttributeSet?) :
    LinearLayout(context, attributeSet) {

    private var mObserver: DataSetObserver? = null
    var adapter: BaseAdapter? = null
        set(value) {
            if (field === value) {
                return
            }
            if (mObserver == null) {
                mObserver = LinearAdapterLayoutObserver(this)
            }
            val baseAdapter2 = field
            baseAdapter2?.unregisterDataSetObserver(mObserver)
            field = value
            val baseAdapter3 = field
            baseAdapter3?.registerDataSetObserver(mObserver)
            updateChildView()
        }

    fun notifyDataSetChange() {
        updateChildView()
    }

    fun updateChildView() {
        adapter?.let { adapter ->
            val arrayList = ArrayList<View?>()
            for (i in 0 until adapter.count) {
                arrayList.add(adapter.getView(i, getChildAt(i), this))
            }
            removeAllViewsInLayout()
            for (i2 in arrayList.indices) {
                var layoutParams = (arrayList[i2] as View).layoutParams
                if (layoutParams == null) {
                    layoutParams = generateDefaultLayoutParams()
                }
                addViewInLayout(arrayList[i2] as View, i2, layoutParams)
            }
            requestLayout()
            invalidate()
        } ?: removeAllViews()
    }

    internal class LinearAdapterLayoutObserver(private val linearAdapterLayout: LinearAdapterLayout) :
        DataSetObserver() {
        override fun onChanged() {
            updateChildView1(linearAdapterLayout)
        }
    }

    companion object {
        fun updateChildView1(linearAdapterLayout: LinearAdapterLayout) {
            linearAdapterLayout.updateChildView()
        }
    }
}
